kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/OperatorMain.java

```diff
@@ -7,47 +7,100 @@
 
 import java.time.Duration;
 import java.util.List;
+import java.util.function.Supplier;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import io.fabric8.kubernetes.client.KubernetesClient;
 import io.javaoperatorsdk.operator.Operator;
+import io.javaoperatorsdk.operator.monitoring.micrometer.MicrometerMetrics;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Metrics;
+import io.micrometer.core.instrument.composite.CompositeMeterRegistry;
+import io.micrometer.prometheusmetrics.PrometheusConfig;
+import io.micrometer.prometheusmetrics.PrometheusMeterRegistry;
 
 import io.kroxylicious.kubernetes.operator.config.FilterApiDecl;
 import io.kroxylicious.kubernetes.operator.config.RuntimeDecl;
+import io.kroxylicious.proxy.tag.VisibleForTesting;
 
 import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
 
 /**
  * The {@code main} method entrypoint for the operator
  */
 public class OperatorMain {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(OperatorMain.class);
+    private MeterRegistry registry;
+    private final Operator operator;
+    private final Supplier<CompositeMeterRegistry> globalRegistrySupplier;
+
+    public OperatorMain(Supplier<CompositeMeterRegistry> globalRegistrySupplier) {
+        this(globalRegistrySupplier, null);
+    }
+
+    @VisibleForTesting
+    OperatorMain(Supplier<CompositeMeterRegistry> globalRegistrySupplier, @Nullable KubernetesClient client) {
+        this.globalRegistrySupplier = globalRegistrySupplier;
+        final MicrometerMetrics metrics = enablePrometheusMetrics();
+        // o.withMetrics is invoked multiple times so can cause issues with enabling metrics.
+        operator = new Operator(o -> {
+            o.withMetrics(metrics);
+            if (client != null) {
+                o.withKubernetesClient(client);
+            }
+        });
+    }
 
     public static void main(String[] args) {
         try {
-            run();
+            new OperatorMain(() -> Metrics.globalRegistry).run();
         }
         catch (Exception e) {
             LOGGER.error("Operator has thrown exception during startup. Will now exit.", e);
             System.exit(1);
         }
     }
 
-    static void run() {
-        Operator operator = new Operator();
+    void run() {
```

Comment:
* What's the benefit of having the starting of the operator as separate method, rather than doing it in the constructor?
* If we do see a benefit, perhaps we rename this method to `start`? 
* Let's add some doc to explain that this method returns after the operator has successfully started. 
* Should we do anything special if you try to start/run the same operator multiple times?
* Should you be able to restart a stopped operator?

Reply:
>What's the benefit of having the starting of the operator as separate method, rather than doing it in the constructor?

Yes. Starting it in the ctor makes unit testing an utter nightmare. Separating the two makes it a little easier.

> If we do see a benefit, perhaps we rename this method to start?
> Let's add some doc to explain that this method returns after the operator has successfully started.

Happy to, I was trying to do as little as possible to keep the review small.

> Should we do anything special if you try to start/run the same operator multiple times?

Shrug, depends how defensive we want to be. I don't see a lot of value in preventing it.

> Should you be able to restart a stopped operator?

No.